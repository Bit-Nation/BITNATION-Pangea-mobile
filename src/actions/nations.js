// @flow

import type { NationIdType, NationType } from '../types/Nation';

export type NationTab = 'ALL_NATIONS' | 'MY_NATIONS';

type SwitchNationTabAction = { +type: 'SWITCH_NATIONS_TAB', +tab: NationTab };
type OpenNationAction = { +type: 'OPEN_NATION', +nationId: NationIdType };
type RequestSyncNationsAction = { +type: 'START_NATIONS_SYNC' };
type JoinNationAction = { +type: 'REQUEST_JOIN_NATION' };
type LeaveNationAction = { +type: 'REQUEST_LEAVE_NATION' };
type FetchNationsStartedAction = { +type: 'NATIONS_FETCH_STARTED' };
type DoneFetchNationsAction = { +type: 'DONE_FETCH_NATIONS' };
type DoneSyncNationsAction = { +type: 'DONE_SYNC_NATIONS', +payload: Array<NationType> };

export type Action =
  | SwitchNationTabAction
  | OpenNationAction
  | FetchNationsStartedAction
  | RequestSyncNationsAction
  | JoinNationAction
  | LeaveNationAction
  | DoneFetchNationsAction
  | DoneSyncNationsAction;

export const SWITCH_NATIONS_TAB = 'SWITCH_NATIONS_TAB';
export const OPEN_NATION = 'OPEN_NATION';
export const START_NATIONS_SYNC = 'START_NATIONS_SYNC';
export const NATIONS_FETCH_STARTED = 'NATIONS_FETCH_STARTED';
export const DONE_FETCH_NATIONS = 'DONE_FETCH_NATIONS';
export const CANCEL_LOADING = 'CANCEL_LOADING';
export const REQUEST_JOIN_NATION = 'REQUEST_JOIN_NATION';
export const REQUEST_LEAVE_NATION = 'REQUEST_LEAVE_NATION';
export const DONE_SYNC_NATIONS = 'DONE_SYNC_NATIONS';

/**
 * @desc Action creator for an action that switches nation list filter.
 * @param {NationTab} tab Tab to switch to.
 * @returns {SwitchNationTabAction} An action.
 */
export function switchNationTab(tab: NationTab): SwitchNationTabAction {
  return {
    type: SWITCH_NATIONS_TAB,
    tab,
  };
}

/**
 * @desc Action creator for an action that should be called before show nation details screen.
 * @param {NationIdType} id Id of nation to open.
 * @returns {OpenNationAction} An action.
 */
export function openNation(id: NationIdType): OpenNationAction {
  return {
    type: OPEN_NATION,
    nationId: id,
  };
}

/**
 * @desc Action creator for an action that is called when nations fetch started.
 * @returns {FetchNationsStartedAction} An action.
 */
export function fetchNationsStarted(): FetchNationsStartedAction {
  return {
    type: NATIONS_FETCH_STARTED,
  };
}

/**
 * @desc Action creator for an action that starts currently opened nation join.
 * @returns {JoinNationAction} An action.
 */
export function joinNation(): JoinNationAction {
  return {
    type: REQUEST_JOIN_NATION,
  };
}

/**
 * @desc Action creator for an action that starts currently opened nation leave.
 * @returns {LeaveNationAction} An action.
 */
export function leaveNation(): LeaveNationAction {
  return {
    type: REQUEST_LEAVE_NATION,
  };
}

/**
 * @desc Action creator for an action that is called when nations fetch finished.
 * @returns {DoneFetchNationsAction} An action
 */
export function doneFetchNations(): DoneFetchNationsAction {
  return {
    type: DONE_FETCH_NATIONS,
  };
}

/**
 * @desc Action creator for an action that is called when nations sync finished.
 * @param {NationType[]} nations Nations coming from database.
 * @returns {DoneSyncNationsAction} An action.
 */
export function doneSyncNations(nations: Array<NationType>): DoneSyncNationsAction {
  return {
    type: DONE_SYNC_NATIONS,
    payload: nations,
  };
}
