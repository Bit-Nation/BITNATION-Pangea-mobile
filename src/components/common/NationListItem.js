// @flow

import React from 'react';
import { Text } from 'react-native';
import { MediaQueryStyleSheet } from 'react-native-responsive';

import GlobalStyles from '../../global/Styles';
import Colors from '../../global/colors';
import ListItem from './ListItem';

type Props = {
  /**
   * @desc Text to display on item
   * @type string
   */
  nationName: string,
  /**
   * @desc Style object for basic text style
   * @type object
   */
  textStyle?: any,
  /**
   * @desc Status of the Nation to display on item
   * @type string
   */
  status?: string,
  /**
   * @desc Id that will be passed in onPress callback.
   * @type string
   */
  id: any,
  /**
   * @desc Callback on press item.
   * @type string
   */
  onPress: (any) => void,
  /**
   * @desc Color Status of the Nation's label to display on item
   * @type string
   */
  statusColor?: string,
}

/**
 * @desc Component that renders nations list item.
 * @return {React.Component} A component.
 */
const NationListItem = ({
  id, textStyle, onPress, nationName, status, statusColor,
}: Props) => {
  const styles = MediaQueryStyleSheet.create({
    ...GlobalStyles,
  });

  return (
    <ListItem
      text={nationName}
      textStyle={textStyle}
      id={id}
      onPress={onPress}
      additionalView={() => (
        <Text style={[styles.listItemTextState, { color: statusColor }]}>
          {status}
        </Text>
      )}
    />
  );
};

NationListItem.defaultProps = {
  statusColor: Colors.listItemTextState.default,
  textStyle: undefined,
  status: undefined,
  id: null,
  onPress: () => undefined,
};

export default NationListItem;
