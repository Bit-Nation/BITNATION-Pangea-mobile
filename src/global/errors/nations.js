export class NationAlreadySubmitted extends Error {
  transKey: string = 'nation.alreadySubmitted';
}

export class StateMutateNotPossible extends Error {
  transKey: string = 'nation.stateMutateNotPossible';
}
