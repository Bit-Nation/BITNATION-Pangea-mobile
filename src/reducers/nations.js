// @flow
import _ from 'lodash';

import {
  type Action,
  type NationTab,
  SWITCH_NATIONS_TAB,
  OPEN_NATION,
  DONE_FETCH_NATIONS,
  CANCEL_LOADING,
  NATIONS_FETCH_STARTED,
  REQUEST_JOIN_NATION,
  REQUEST_LEAVE_NATION,
  NATIONS_UPDATED,
} from '../actions/nations';
import type { NationType, NationIdType, EditingNationType } from '../types/Nation';
import { resolveNation } from '../utils/nations';

export type State = {
  +nations: Array<NationType>,
  +myNationIds: Array<NationIdType>,
  +searchString: string | null,
  +selectedTab: NationTab,
  +openedNationId: NationIdType | null,
  +creatingNation: EditingNationType | null,
  +inProgress: boolean,
}

export const initialState: State = {
  nations: [],
  myNationIds: [],
  searchString: null,
  selectedTab: 'ALL_NATIONS',
  openedNationId: null,
  creatingNation: null,
  inProgress: false,
};

/**
 * @desc Nations reducer.
 * @param {State} state Current state.
 * @param {Action} action Performed action.
 * @returns {State} Next state.
 */
export default (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case SWITCH_NATIONS_TAB:
      return {
        ...state,
        selectedTab: action.tab,
      };
    case OPEN_NATION:
      return {
        ...state,
        openedNationId: action.nationId,
      };
    case NATIONS_FETCH_STARTED:
      return {
        ...state,
        inProgress: true,
      };
    case NATIONS_UPDATED: {
      const myNationIds = _(action.nations)
        .filter(nation => nation.joined)
        .map(nation => nation.id)
        .value();
      return {
        ...state,
        nations: action.nations,
        myNationIds,
      };
    }
    case DONE_FETCH_NATIONS:
      return {
        ...state,
        inProgress: false,
      };
    case REQUEST_JOIN_NATION:
      return {
        ...state,
        inProgress: true,
      };
    case REQUEST_LEAVE_NATION:
      return {
        ...state,
        inProgress: true,
      };
    case CANCEL_LOADING:
      return {
        ...state,
        inProgress: false,
      };
    default:
      return state;
  }
};

export const openedNation = (state: State): NationType | null => {
  if (state.openedNationId !== null) {
    return resolveNation(state.nations, state.openedNationId);
  }
  return null;
};
