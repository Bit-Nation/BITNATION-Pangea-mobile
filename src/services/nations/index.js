// @flow

import Realm from 'realm';
import { Thread } from 'react-native-threads';
import EthereumService from '../ethereum';
import type { NationType, EditingNationType, DBNationType, NationIdType } from '../../types/Nation';
import { convertDraftToDatabase, convertNationToBlockchain } from '../../utils/nations';
import { NationAlreadySubmitted, StateMutateNotPossible } from '../../global/errors/nations';
import { DatabaseWriteFailed } from '../../global/errors/common';
import { jobFactory } from '../txProcessor';
import {
  NATION_DEV_CONTRACT_CREATION_BLOCK, NATION_PROD_CONTRACT_CREATION_BLOCK, TX_JOB_STATUS,
  TX_JOB_TYPE,
} from '../../global/Constants';

export default class NationsService {
  constructor(ethereumService: EthereumService, dbPromise: Promise<Realm>, accountId: string) {
    this.ethereumService = ethereumService;
    this.dbPromise = dbPromise;
    this.currentAccountId = accountId;
    this.logsProcessingPromise = Promise.resolve();
  }

  ethereumService: EthereumService;
  dbPromise: Promise<Realm>;
  currentAccountId: string;
  logsProcessingPromise: Promise<void>;

  // Drafts operations

  async updateDraft(nationId: number, nationData: EditingNationType): Promise<DBNationType> {
    const db = await this.dbPromise;
    const oldNation = await this.nationById(nationId);

    // idInSmartContract >= 0 only when the nation was written to the blockchain
    // @todo we really really need to check for if the nation was already submitted. But we need to have the tx queue thing.
    if (oldNation.idInSmartContract >= 0) {
      throw new NationAlreadySubmitted();
    }

    try {
      let nation: DBNationType;
      db.write(() => {
        nation = db.create('Nation', convertDraftToDatabase(nationData, nationId, this.currentAccountId), true);
      });

      return (nation: any);
    } catch (error) {
      throw new DatabaseWriteFailed();
    }
  }

  async saveDraft(nationData: EditingNationType): Promise<DBNationType> {
    const db = await this.dbPromise;
    try {
      let nation: DBNationType;
      const newId = await this.newNationId();
      db.write(() => {
        nation = db.create('Nation', convertDraftToDatabase(nationData, newId, this.currentAccountId));
      });

      return (nation: any);
    } catch (error) {
      throw new DatabaseWriteFailed(error);
    }
  }

  async submitDraft(nationId: number): Promise<DBNationType> {
    const db = await this.dbPromise;
    const nation = await this.nationById(nationId);

    // idInSmartContract >= 0 only when the nation was written to the blockchain
    // @todo we really really need to check for if the nation was already submitted. But we need to have the tx queue thing.
    if (nation.idInSmartContract >= 0) {
      throw new NationAlreadySubmitted();
    }

    if (this.canMutateNationState(nation) === false) {
      throw new StateMutateNotPossible();
    }

    const nationData = convertNationToBlockchain(nation);
    const tx = await this.ethereumService.nations.createNation(JSON.stringify(nationData));

    const txJob = await jobFactory(tx.hash, TX_JOB_TYPE.NATION_CREATE, this.currentAccountId);
    try {
      db.write(() => {
        nation.tx = txJob;
      });

      return nation;
    } catch (error) {
      throw new DatabaseWriteFailed(error);
    }
  }

  async saveAndSubmit(nationData: EditingNationType): Promise<DBNationType> {
    const draft = await this.saveDraft(nationData);
    return this.submitDraft(draft.id);
  }

  async deleteDraft(nationId: number): Promise<void> {
    const db = await this.dbPromise;
    const nation = await this.nationById(nationId);

    if (nation.idInSmartContract >= 0) {
      throw new NationAlreadySubmitted();
    }

    try {
      db.write(() => {
        db.delete(nation);
      });
    } catch (error) {
      throw new DatabaseWriteFailed(error);
    }
  }

  // Join/leave

  async joinNation(nationData: NationType): Promise<void> {
    const db = await this.dbPromise;
    const nation = await this.nationById(nationData.id);

    if (this.canMutateNationState(nation) === false) {
      throw new StateMutateNotPossible();
    }

    const tx = await this.ethereumService.nations.joinNation(nation.idInSmartContract);
    const txJob = await jobFactory(tx.hash, TX_JOB_TYPE.NATION_JOIN, this.currentAccountId);
    try {
      db.write(() => {
        nation.tx = txJob;
      });
    } catch (_) {
      throw new DatabaseWriteFailed();
    }
  }

  async leaveNation(nationData: NationType): Promise<void> {
    const db = await this.dbPromise;
    const nation = await this.nationById(nationData.id);

    if (this.canMutateNationState(nation) === false) {
      throw new StateMutateNotPossible();
    }

    const tx = await this.ethereumService.nations.leaveNation(nation.idInSmartContract);
    const txJob = await jobFactory(tx.hash, TX_JOB_TYPE.NATION_LEAVE, this.currentAccountId);
    try {
      db.write(() => {
        nation.tx = txJob;
      });
    } catch (_) {
      throw new DatabaseWriteFailed();
    }
  }

  async registerNationIndexing() {
    const firstBlock = this.ethereumService.network === 'dev' ? NATION_DEV_CONTRACT_CREATION_BLOCK : NATION_PROD_CONTRACT_CREATION_BLOCK;

    return new Promise(async (resolve, reject) => {
      const self = this;
      let expectedNationsNumber = (await this.ethereumService.nations.numNations()).toNumber();

      this.ethereumService.nations.onnationcreated = function processLog() {
        // BE CAREFUL! Since strange API of ether.js log passed here as a 'this'.
        const log = this;

        self.performNationUpdate(log.args.nationId.toNumber(), log.transactionHash)
          .then(() => {
            expectedNationsNumber -= 1;
            if (expectedNationsNumber === 0) {
              resolve();
            }
          })
          .catch((error) => {
            console.log(error.toString());
            reject(error);
          });
      };

      if (expectedNationsNumber === 0) {
        resolve();
      }

      this.ethereumService.nations.provider.resetEventsBlock(firstBlock);
    });
  }

  async performNationUpdate(idInSmartContract: number, txHash: string | null) {
    const db = await this.dbPromise;
    const draftToUpdate: DBNationType = db.objects('Nation').filtered(`tx.txHash = '${txHash || ''}' AND tx.type = '${TX_JOB_TYPE.NATION_CREATE}'`)[0];
    if (draftToUpdate != null && draftToUpdate.idInSmartContract !== idInSmartContract) {
      // It's a draft that we need to update in database as submitted nation
      db.write(() => {
        draftToUpdate.idInSmartContract = idInSmartContract;
      });
    }

    // For some reason we sometimes get object instead of array here. This object contains nations that we don't actually join. So we ignore it.
    const joinedNations = (await this.ethereumService.nations.getJoinedNations({ from: this.ethereumService.wallet.address }));
    let isNationJoined: boolean = false;
    if (Array.isArray(joinedNations) === true) {
      isNationJoined = joinedNations.map(bigNumber => bigNumber.toNumber()).includes(idInSmartContract);
    }
    const citizensNumber = (await this.ethereumService.nations.getNumCitizens(idInSmartContract)).toNumber();

    const nationToUpdate: DBNationType = db.objects('Nation').filtered(`accountId = '${this.currentAccountId}' && idInSmartContract = ${idInSmartContract}`)[0];
    if (nationToUpdate != null && (nationToUpdate.joined !== isNationJoined || nationToUpdate.citizens !== citizensNumber)) {
      // It's a nation that somehow is already in database, so we just update it.
      db.write(() => {
        nationToUpdate.joined = isNationJoined;
        nationToUpdate.citizens = citizensNumber;
      });
    }

    if (draftToUpdate != null || nationToUpdate != null) {
      // Since we update a nation we don't need to create it again.
      return;
    }

    // const nationData = JSON.parse(await this.ethereumService.nations.getNationMetaData(idInSmartContract));
    const nationDataJson = JSON.parse(await this.ethereumService.nations.getNationMetaData(idInSmartContract));
    nationDataJson.idInSmartContract = idInSmartContract;
    nationDataJson.joined = isNationJoined;
    nationDataJson.citizens = citizensNumber;
    nationDataJson.accountId = this.currentAccountId;
    nationDataJson.id = await this.newNationId();
    const thread = new Thread('worker.js');
    thread.postMessage(JSON.stringify(nationDataJson));
    thread.onmessage = (message) => { console.log(message); };
    /* const newId = await this.newNationId();
    db.write(() => {
      db.create('Nation', {
        id: newId,
        accountId: this.currentAccountId,
        idInSmartContract,
        nationName: nationData.nationName,
        nationDescription: nationData.nationDescription,
        created: true,
        exists: nationData.exists,
        virtualNation: nationData.virtualNation,
        nationCode: nationData.nationCode,
        lawEnforcementMechanism: nationData.lawEnforcementMechanism,
        profit: nationData.profit,
        nonCitizenUse: nationData.nonCitizenUse,
        diplomaticRecognition: nationData.diplomaticRecognition,
        decisionMakingProcess: nationData.decisionMakingProcess,
        governanceService: nationData.governanceService,
        joined: isNationJoined,
        citizens: citizensNumber,
      });
    }); */
  }

  // Utilities

  async newNationId(): Promise<number> {
    const db = await this.dbPromise;
    const sorted = db.objects('Nation').sorted('id', true);
    return (sorted.length === 0 ? 1 : sorted[0].id + 1);
  }

  async nationById(id: NationIdType): Promise<DBNationType> {
    const db = await this.dbPromise;
    const nations = db.objects('Nation').filtered(`id = ${id}`);
    if (nations.length === 0) {
      throw new Error('system_error.nation.does_not_exist');
    }

    return nations[0];
  }

  cleanUp() {
    this.ethereumService.nations.onnationcreated = null;
  }

  /**
   * @desc Check if the blockchain state of the nation can be mutated
   * @param {DBNationType | NationType} nation Nation to check
   * @returns {boolean} Return true if state can be mutated
   */
  canMutateNationState = (nation: DBNationType | NationType): boolean => {
    if (nation.tx == null) {
      return true;
    }
    return nation.tx.status !== TX_JOB_STATUS.PENDING;
  };
}
