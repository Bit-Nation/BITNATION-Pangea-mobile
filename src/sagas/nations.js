import { call, put, select, takeEvery } from 'redux-saga/effects';
import { Alert } from 'react-native';

import {
  CANCEL_NATION_CREATE, DONE_NATION_CREATE, START_NATIONS_FETCH, DONE_FETCH_NATIONS,
  NATION_CREATE, CANCEL_LOADING, REQUEST_JOIN_NATION, REQUEST_LEAVE_NATION,
} from '../actions/nations';
import { getPangeaLibrary } from '../services/container';
import { waitConnect } from '../utils/connectivity';
import { CONNECTION_TIMEOUT } from '../global/Constants';
import { resolveNation } from '../utils/nations';

export async function checkConnection() {
  return await waitConnect(CONNECTION_TIMEOUT);
}

const extractMessage = (error) => {
  if (error.toString().indexOf('insufficient') !== -1) {
    return 'Insufficient funds. Please check your wallet';
  }
  return error.toString();
};

export function* createNation(action) {
  if (action.payload) {
    try {
      let pangeaLib = yield call(getPangeaLibrary);
      yield call(checkConnection);
      let result = yield call(pangeaLib.eth.nation.create, action.payload);
      yield put({ type: DONE_NATION_CREATE });
      yield call([action.navigator, 'dismissModal']);
      yield put({ type: START_NATIONS_FETCH });
    } catch (e) {
      console.log('Create nation error: ', e);
      Alert.alert(extractMessage(e));
      yield put({ type: CANCEL_LOADING });
    }
  }
}

export function* fetchNations() {
  try {
    let pangeaLib = yield call(getPangeaLibrary);
    yield call(checkConnection);
    yield call(pangeaLib.eth.nation.index);
    const updatedNations = yield call(pangeaLib.eth.nation.all);
    yield put({ type: DONE_FETCH_NATIONS, payload: [...updatedNations] });
  } catch (e) {
    console.log('Update nation error: ', e);
    Alert.alert(extractMessage(e));
    yield put({ type: CANCEL_LOADING });
  }
}

export const getNations = state => state.nations

export function* joinNation() {
  try {
    let pangeaLib = yield call(getPangeaLibrary);
    let nationsState = yield select(getNations);
    const currentNation = resolveNation(nationsState.nations, nationsState.openedNationId);
    yield call(checkConnection);
    let result = yield call(pangeaLib.eth.nation.joinNation, currentNation.id);
    // console.log('joined nation: ', result);
    yield put({ type: CANCEL_LOADING });
    yield put({ type: START_NATIONS_FETCH });
  } catch (e) {
    console.log('Join nation error: ', e);
    Alert.alert(extractMessage(e));
    yield put({ type: CANCEL_LOADING });
  }
}

export function* leaveNation() {
  try {
    let pangeaLib = yield call(getPangeaLibrary);
    let nationsState = yield select(getNations);
    const currentNation = resolveNation(nationsState.nations, nationsState.openedNationId);
    yield call(checkConnection);
    let result = yield call(pangeaLib.eth.nation.leaveNation, currentNation.id);
    // console.log('leave nation: ', result);
    yield put({ type: CANCEL_LOADING });
    yield put({ type: START_NATIONS_FETCH });
  } catch (e) {
    console.log('Leave nation error: ', e);
    Alert.alert(extractMessage(e));
    yield put({ type: CANCEL_LOADING });
  }
}

export default function* watchProfileUpdate() {
  yield takeEvery(NATION_CREATE, createNation);
  yield takeEvery(START_NATIONS_FETCH, fetchNations);
  yield takeEvery(REQUEST_JOIN_NATION, joinNation);
  yield takeEvery(REQUEST_LEAVE_NATION, leaveNation);
}
