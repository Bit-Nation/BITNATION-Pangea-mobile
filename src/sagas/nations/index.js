// @flow

import { all, takeEvery, call } from 'redux-saga/effects';

import { REQUEST_JOIN_NATION, REQUEST_LEAVE_NATION } from '../../actions/nations';
import { joinNation, leaveNation, startDatabaseListening, startNationIndexingWorker } from './sagas';
import { SERVICES_CREATED } from '../../actions/serviceContainer';

/**
 * @desc Root modify nation saga.
 * @return {void}
 */
export default function* rootSaga(): Generator<*, *, *> {
  yield all([
    yield takeEvery(REQUEST_JOIN_NATION, joinNation),
    yield takeEvery(REQUEST_LEAVE_NATION, leaveNation),
    yield takeEvery(SERVICES_CREATED, startNationIndexingWorker),
    yield call(startDatabaseListening),
  ]);
}
