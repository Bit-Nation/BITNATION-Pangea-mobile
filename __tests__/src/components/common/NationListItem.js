import React from 'react';
import { shallow } from 'enzyme';

import NationListItem from '../../../../src/components/common/NationListItem';
import { statusColor } from '../../../../src/utils/nations';

describe('NationListItem tests', () => {
  describe('Rendering', () => {
    test('With text', () => {
      const wrapper = shallow(<NationListItem text='Test list item' />);
      expect(wrapper).toMatchSnapshot();
    });

    test('With text colors', () => {
      const wrapper = shallow(<NationListItem
        text='Test list item'
        statusColor={statusColor(0)}
      />);
      expect(wrapper).toMatchSnapshot();
    });

    test('Default has onPress', () => {
      expect(NationListItem.defaultProps.onPress).toBeDefined();
    });
  });

  describe('Behaviour', () => {
    test('Press', () => {
      const mockFunc = jest.fn();
      const id = 'Test id';
      const wrapper = shallow(<NationListItem text='Test list item' onPress={mockFunc} id={id} />);
      expect(wrapper).toMatchSnapshot();
      const render = wrapper.dive();
      const touchables = render.find('[testIDJest="Touchable"]');
      expect(touchables).toHaveLength(1);
      touchables.props().onPress();
      expect(mockFunc).toHaveBeenCalledTimes(1);
      expect(mockFunc).toHaveBeenCalledWith(id);
    });
  });
});
