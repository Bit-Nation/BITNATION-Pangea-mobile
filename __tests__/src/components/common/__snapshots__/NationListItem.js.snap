// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NationListItem tests Behaviour Press 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NationListItem
    id="Test id"
    onPress={[MockFunction]}
    status={undefined}
    statusColor="gray"
    text="Test list item"
    textStyle={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "AdditionalView": [Function],
      "additionalView": [Function],
      "disabled": false,
      "disclosureIconVisible": true,
      "iconSource": undefined,
      "id": "Test id",
      "onPress": [MockFunction],
      "subtitle": undefined,
      "text": undefined,
      "textStyle": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "AdditionalView": [Function],
        "additionalView": [Function],
        "disabled": false,
        "disclosureIconVisible": true,
        "iconSource": undefined,
        "id": "Test id",
        "onPress": [MockFunction],
        "subtitle": undefined,
        "text": undefined,
        "textStyle": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`NationListItem tests Rendering With text 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NationListItem
    id={null}
    onPress={[Function]}
    status={undefined}
    statusColor="gray"
    text="Test list item"
    textStyle={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "AdditionalView": [Function],
      "additionalView": [Function],
      "disabled": false,
      "disclosureIconVisible": true,
      "iconSource": undefined,
      "id": null,
      "onPress": [Function],
      "subtitle": undefined,
      "text": undefined,
      "textStyle": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "AdditionalView": [Function],
        "additionalView": [Function],
        "disabled": false,
        "disclosureIconVisible": true,
        "iconSource": undefined,
        "id": null,
        "onPress": [Function],
        "subtitle": undefined,
        "text": undefined,
        "textStyle": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`NationListItem tests Rendering With text colors 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NationListItem
    id={null}
    onPress={[Function]}
    status={undefined}
    statusColor="gray"
    text="Test list item"
    textStyle={undefined}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "AdditionalView": [Function],
      "additionalView": [Function],
      "disabled": false,
      "disclosureIconVisible": true,
      "iconSource": undefined,
      "id": null,
      "onPress": [Function],
      "subtitle": undefined,
      "text": undefined,
      "textStyle": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "AdditionalView": [Function],
        "additionalView": [Function],
        "disabled": false,
        "disclosureIconVisible": true,
        "iconSource": undefined,
        "id": null,
        "onPress": [Function],
        "subtitle": undefined,
        "text": undefined,
        "textStyle": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
